src/effects/backends/builtin/compressoreffect.cpp

```diff
@@ -0,0 +1,236 @@
+    gain->setId("gain");
+    gain->setName(QObject::tr("Output gain"));
+    gain->setShortName(QObject::tr("Gain"));
+    gain->setDescription(
+            QObject::tr("The Output gain knob adjusts the level of the output "
+                        "signal after the compression was applied"));
+    gain->setValueScaler(EffectManifestParameter::ValueScaler::Linear);
+    gain->setUnitsHint(EffectManifestParameter::UnitsHint::Decibel);
+    gain->setRange(-25, 0, 25);
+
+    return pManifest;
+}
+
+CompressorGroupState::CompressorGroupState(
+        const mixxx::EngineParameters& engineParameters)
+        : EffectState(engineParameters),
+          previousStateDB(0),
+          previousMakeUpGain(0) {
+}
+
+void CompressorEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pThreshold = parameters.value("threshold");
+    m_pRatio = parameters.value("ratio");
+    m_pKnee = parameters.value("knee");
+    m_pAttack = parameters.value("attack");
+    m_pRelease = parameters.value("release");
+    m_pGain = parameters.value("gain");
+    m_pAutoMakeUp = parameters.value("automakeup");
+    m_pClipping = parameters.value("clipping");
+}
+
+void CompressorEffect::processChannel(
+        CompressorGroupState* pState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(enableState);
+
+    SINT numSamples = engineParameters.samplesPerBuffer();
+
+    // Compression
+    applyCompression(pState, engineParameters, pInput, pOutput);
+
+    // Auto make up
+    if (m_pAutoMakeUp->toInt() == static_cast<int>(AutoMakeUp::AutoMakeUpOn)) {
+        applyAutoMakeUp(pState, pOutput, numSamples);
+    }
+
+    // Output gain
+    CSAMPLE gainParamDB = static_cast<CSAMPLE>(m_pGain->value());
+    SampleUtil::applyGain(pOutput, db2ratio(gainParamDB), numSamples);
+
+    // Clipping
+    if (m_pClipping->toInt() == static_cast<int>(Clipping::ClippingOn)) {
+        SampleUtil::applyClamp(pOutput, numSamples);
+    }
+}
+
+void CompressorEffect::applyAutoMakeUp(CompressorGroupState* pState,
+        CSAMPLE* pOutput,
+        const SINT& numSamples) {
+    CSAMPLE makeUpStateDB = pState->previousMakeUpGain;
+    CSAMPLE maxSample = SampleUtil::maxAbsAmplitude(pOutput, numSamples);
+    if (maxSample > 0) {
+        CSAMPLE maxSampleDB = ratio2db(maxSample);
+        CSAMPLE minGainReductionDB = -maxSampleDB + kMakeUpTarget;
+        makeUpStateDB = kMakeUpAttackCoeff * minGainReductionDB +
+                (1 - kMakeUpAttackCoeff) * makeUpStateDB;
+        CSAMPLE levelDB = makeUpStateDB + maxSampleDB;
+        // logarithmic smoothing
+        if (levelDB > -1.0) {
+            makeUpStateDB = log10(levelDB + 2.0f) - 1.0f - maxSampleDB;
```

Comment:
Please also take care of the warnings.

Reply:
These peaks still cause harm in your speakers. It is caused by the sub based introduced by the flatened sine waves.

Reply:
I think this message belongs to discussion of clipping? I've already removed clipping :)

Also I've changed float variables to doubles, so I believe it fixes warnings for Ubuntu.

Reply:
Oh yes, thank you.